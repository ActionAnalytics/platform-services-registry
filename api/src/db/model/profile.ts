import { logger } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import { CommonFields, Model } from './model';

export interface ProjectProfile extends CommonFields {
  name: string,
  description: string,
  busOrgId: number,
  active?: boolean,
  prioritySystem?: boolean,
  criticalSystem?: boolean,
}

export default class ProfileModel extends Model {
  table: string = 'profile';
  requiredFields: string[] = [
    'name',
    'description',
    'busOrgId',
    'prioritySystem',
  ];
  pool: Pool;

  constructor(pool: any) {
    super();
    this.pool = pool;
  }

  async create(data: ProjectProfile): Promise<ProjectProfile> {
    const query = {
      text: `
        INSERT INTO ${this.table}
          (name, description, bus_org_id, priority_system, critical_system)
          VALUES ($1, $2, $3, $4, $5) RETURNING *;`,
      values: [
        data.name,
        data.description,
        data.busOrgId,
        data.prioritySystem ? data.prioritySystem : false,
        data.criticalSystem ? data.criticalSystem : false,
      ],
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to create project profile`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async update(profileId, data: ProjectProfile): Promise<ProjectProfile> {
    const values: any[] = [];
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            name = $1, description = $2, bus_org_id = $3,
            active = $4, priority_system = $5, critical_system = $6
          WHERE id = ${profileId}
          RETURNING *;`,
      values,
    };

    try {
      const record = await this.findById(profileId);
      const aData = { ...record, ...data };
      query.values = [
        aData.name,
        aData.description,
        aData.busOrgId,
        aData.active,
        aData.criticalSystem,
        aData.prioritySystem,
      ];

      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to create project profile`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  };

  async delete(profileId): Promise<ProjectProfile> {
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            archived = true
          WHERE id = ${profileId}
          RETURNING *;
      `,
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to archive project profile`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  };
}
