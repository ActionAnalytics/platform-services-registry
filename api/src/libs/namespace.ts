//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Request, Response } from 'express';
import { difference } from 'lodash';
import config from '../config';
import DataManager from '../db';

export const createNamespace = async (req: Request, res: Response): Promise<void> => {
  const dm = new DataManager(config);
  const { NamespaceModel } = dm;
  const { profileId } = req.params;
  const { body } = req;
  const aBody = { ...body, profileId };

  const diff = difference(NamespaceModel.requiredFields, Object.keys(aBody));
  if (diff.length !== 0) {
    throw errorWithCode(`Missing required properties: ${diff}`, 400);
  }

  try {
    const results = await NamespaceModel.create(aBody);

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable create new namespace';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespaces = async (req: Request, res: Response): Promise<void> => {
  const dm = new DataManager(config);
  const { NamespaceModel } = dm;
  const { profileId } = req.params;

  try {
    const results = await NamespaceModel.findForProfile(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileNamespace = async (req: Request, res: Response): Promise<void> => {
  const dm = new DataManager(config);
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = req.params;
  const { body } = req;

  // Make sure these are not updated!
  delete body.id;
  delete body.archived;
  delete body.createdAt;
  delete body.updatedAt;

  const aBody = { ...body, profileId };
  const diff = difference(NamespaceModel.requiredFields, Object.keys(aBody));
  if (diff.length !== 0) {
    throw errorWithCode(`Missing required properties: ${diff}`, 400);
  }

  try {
    const results = await NamespaceModel.update(namespaceId, aBody);

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProfileNamespace = async (req: Request, res: Response): Promise<void> => {
  const dm = new DataManager(config);
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = req.params;

  try {
    await NamespaceModel.delete(namespaceId);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to archive profile ${profileId} namespace ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
