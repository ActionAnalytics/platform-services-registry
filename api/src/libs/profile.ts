//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Request, Response } from 'express';
import config from '../config';
import DataManager from '../db';

export const fetchAllProjectProfiles = async (req: Request, res: Response): Promise<void> => {
  const dm = new DataManager(config);
  const { Profile } = dm;

  try {
    const results = await Profile.findAll();

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable fetch all project profiles';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProjectProfile = async (req: Request, res: Response): Promise<void> => {
  const dm = new DataManager(config);
  const { Profile } = dm;
  const { profileId } = req.params;

  try {
    const results = await Profile.findById(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch project profile with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
