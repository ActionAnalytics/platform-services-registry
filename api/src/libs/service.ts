//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-2.
//

import { JWTServiceManager } from '@bcgov/common-nodejs-utils';
import axios, { AxiosInstance } from 'axios';
import config from '../config';

export interface Dependency {
  name: string;
  healthy: boolean;
  info: string;
}

export interface SSOConfig {
  uri: string;
  grantType: string;
  clientId: string;
  clientSecret: string;
}

export interface Options {
  sso: SSOConfig;
  baseURL: string;
}

export const enum BodyType {
  HTML = 'html',
  Text = 'text',
}

export const enum Encoding {
  Default = 'utf-8',
  Base64 = 'base64',
  Binary = 'binary',
  Hex = 'hex',
}

export const enum Priority {
  Default = 'normal',
  Low = 'low',
  High = 'high',
}

export const enum AttachmentContentType {
  String = 'string',
}

export interface Attachment {
  content: string;
  contentType: AttachmentContentType.String;
  encoding: Encoding.Base64;
  filename: string;
}

export interface Message {
  attachments?: Attachment[];
  bcc?: string[];
  bodyType: BodyType;
  body: string;
  cc?: string[];
  delayTS?: Date;
  encoding?: Encoding;
  from: string;
  priority?: Priority;
  subject: string;
  tag?: string;
  to: string[];
}

export interface SentMessage {
  to: string[];
  messageId: string;
}

export interface SendReciept {
  transactionId: string;
  messages: SentMessage[];
}

export class CommonEmailService {
  private tokenManager: JWTServiceManager;
  private axi: AxiosInstance;

  constructor(options: Options) {
    this.tokenManager = new JWTServiceManager(options.sso);
    this.axi = axios.create({
      baseURL: options.baseURL,
    });
  }

  public async health(): Promise<Dependency[]> {
    try {
      const token = await this.tokenManager.accessToken;
      const response = await this.axi.get('health', {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
      });

      const { dependencies } = response.data;

      return dependencies
    } catch (err) {
      const errmsg = 'Unable to fetch health status';
      throw new Error(`${errmsg}, reason = ${err.message}`);
    }
  }

  public async send(message: Message) {
    try {
      const token = await this.tokenManager.accessToken;
      const response = await this.axi.post('email', message, {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
      });

      const { txId, messages } = response.data;
      const theMessages = messages.map(m => {
        return {
          to: m.to,
          messageId: m.msgId,
        }
      });

      return {
        transactionId: txId,
        messages: theMessages,
      }
    } catch (err) {
      const errmsg = 'Unable to send message';
      throw new Error(`${errmsg}, reason = ${err.message}`);
    }
  }
}

const main = async () => {
  const options: Options = {
    baseURL: config.get('ches:baseURL'),
    sso: {
      uri: config.get('ches:ssoTokenURL'),
      grantType: config.get('ches:ssoGrantType'),
      clientId: config.get('ches:ssoClientId'),
      clientSecret: config.get('ches:ssoClientSecret'),
    },
  }
  const message: Message = {
    bodyType: BodyType.Text,
    body: 'Hello World',
    from: 'jason.leach@gov.bc.ca',
    subject: 'Test 123',
    to: ['jason.leach@fullboar.ca'],
  }

  const x = new CommonEmailService(options);
  // await x.health();
  const reciept = await x.send(message);
  console.log(reciept);
}

main();
