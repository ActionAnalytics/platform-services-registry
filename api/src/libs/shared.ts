//
// Code Sign
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-07-23.
//

'use strict';

import { JWTServiceManager } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import config from '../config';

interface Shared {
  pgPool: Pool;
}

const ssoKey = Symbol.for('ca.bc.gov.platsrv.sso');
const pgPoolKey = Symbol.for('ca.bc.gov.platsrv.pgpool');
const gs = Object.getOwnPropertySymbols(global);

const main = async () => {
  if (!(gs.indexOf(ssoKey) > -1)) {
    global[ssoKey] = new JWTServiceManager({
      uri: config.get('sso:tokenUrl'),
      grantType: config.get('sso:grantType'),
      clientId: config.get('sso:clientId'),
      clientSecret: config.get('sso:clientSecret'),
    });
  }

  if (!(gs.indexOf(pgPoolKey) > -1)) {
    const params = {
      host: config.get('db:host'),
      port: config.get('db:port'),
      database: config.get('db:database'),
      user: config.get('db:user'),
      password: config.get('db:password'),
      max: 5,
      idleTimeoutMillis: 30000,
      connectionTimeoutMillis: 2000,
    }

    global[pgPoolKey] = new Pool(params);
  }
}

main();

const shared = {};

Object.defineProperty(shared, 'sso', {
  get: () => global[ssoKey],
});

Object.defineProperty(shared, 'pgPool', {
  get: () => global[pgPoolKey],
});

Object.freeze(shared);

export default shared as Shared;
