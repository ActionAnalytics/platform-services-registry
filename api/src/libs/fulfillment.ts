//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-2.
//

import { logger } from '@bcgov/common-nodejs-utils';
import fs from 'fs';
import path from 'path';
import { SUBJECTS } from '../constants';
import DataManager from '../db';
import { Contact } from '../db/model/contact';
import { ProjectProfile } from '../db/model/profile';
import { BodyType, Message } from '../libs/service';
import shared from './shared';

export interface Context {
  something: any;
}

export const provisioningStartedMessage = async (profileId: number): Promise<Message | undefined> => {

  const dm = new DataManager(shared.pgPool);
  const { ContactModel } = dm;
  const buff = fs.readFileSync(path.join(__dirname, '../../', 'templates/provisioning-request-received.txt'));
  const contacts: Contact[] = await ContactModel.findForProject(profileId);
  const to = [...new Set(contacts.map(c => c.email))];

  if (to.length === 0) {
    return;
  }

  const message: Message = {
    bodyType: BodyType.Text,
    body: buff.toString('utf8'),
    to,
    from: 'Registry <pathfinder@gov.bc.ca>',
    subject: 'Namespace Provisioning',
  }

  return message;
}

export const contextForProvisioning = async (profileId: number): Promise<any> => {

  try {
    const dm = new DataManager(shared.pgPool);
    const { ProfileModel, ContactModel, NamespaceModel } = dm;
    const profile: ProjectProfile = await ProfileModel.findById(profileId);
    const contacts: Contact[] = await ContactModel.findForProject(profileId);
    const contact = contacts.filter(c => c.roleId === 2).pop();
    const namespaces = await NamespaceModel.findForProfile(profileId);

    if (!profile || !contact || !namespaces) {
      logger.error('Unable to create context for provisioning');
      return; // This is a problem.
    }

    const context = {
      profileId: profile.id,
      displayName: profile.name,
      namespaces,
      technicalContact: {
        githubId: contact.githubId,
      },
    };

    return context;
  } catch (err) {
    const message = `Unable to build contect for profile ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};

export const fulfillNamespaceProvisioning = async (profileId: number) =>
  new Promise(async (resolve, reject) => {

    try {
      const nc = shared.nats;
      const subject = SUBJECTS.NSPROVISION;
      const context = await contextForProvisioning(profileId);

      nc.on('error', () => {
        const errmsg = `NATS error sending order ${profileId} to ${subject}`;
        reject(new Error(errmsg));
      });

      nc.publish(subject, context);

      nc.flush(() => {
        nc.removeAllListeners(['error']);
      });

      const message = await provisioningStartedMessage(profileId);
      if (message) {
        await shared.ches.send(message);
      }

      resolve();
    } catch (err) {
      const message = `Unable to provision namespaces for profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  });
