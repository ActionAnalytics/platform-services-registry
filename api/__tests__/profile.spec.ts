//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { Response } from 'express';
import fs from 'fs';
import path from 'path';
import { fetchAllProjectProfiles } from '../src/libs/profile';

const p0 = path.join(__dirname, 'fixtures/select-profiles.json');
const selectProfiles = JSON.parse(fs.readFileSync(p0, 'utf8'));

const pmock = {
  connect: jest.fn(),
  query: jest.fn().mockReturnValue({ rows: selectProfiles }),
  end: jest.fn(),
};

jest.mock('pg', () => {
  return { Pool: jest.fn(() => pmock) };
});

class FauxExpress {
  res: Partial<Response> = {
    clearCookie: jest.fn(),
    cookie: jest.fn(),
    json: jest.fn().mockImplementation((param) => {
      this.responseData = param;
      return this.res;
    }),
    status: jest.fn().mockImplementation((code) => {
      this.res.statusCode = code;
      return this.res;
    }),
    statusCode: 200,
  }
  req: any;
  responseData: any;
}


describe('Profile event handlers', () => {
  const ex = new FauxExpress();

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  it('All profiles are returned', async () => {
    // const req = {
    //   params: { profileId: 3 },
    // };
    const req = {};

    // @ts-ignore
    await fetchAllProjectProfiles(req, ex.res);

    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });
});
