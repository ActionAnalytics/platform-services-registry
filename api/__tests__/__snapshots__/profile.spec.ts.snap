// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile event handlers A single profile is returned 1`] = `200`;

exports[`Profile event handlers A single profile is returned 2`] = `
Object {
  "createdAt": "2020-04-28T00:00:00.000Z",
  "description": null,
  "id": 1,
  "name": "Health Gateway",
  "updatedAt": "2020-04-28T00:00:00.000Z",
}
`;

exports[`Profile event handlers All profiles are returned 1`] = `200`;

exports[`Profile event handlers All profiles are returned 2`] = `
Array [
  Object {
    "createdAt": "2020-04-28T00:00:00.000Z",
    "description": null,
    "id": 1,
    "name": "Health Gateway",
    "updatedAt": "2020-04-28T00:00:00.000Z",
  },
  Object {
    "createdAt": "2020-04-28T00:00:00.000Z",
    "description": "Hello World",
    "id": 2,
    "name": "EPIC",
    "updatedAt": "2020-04-28T00:00:00.000Z",
  },
  Object {
    "createdAt": "2020-04-28T00:00:00.000Z",
    "description": "This is some description",
    "id": 3,
    "name": "Mines Digital Services",
    "updatedAt": "2020-04-28T00:00:00.000Z",
  },
  Object {
    "createdAt": "2020-04-28T00:00:00.000Z",
    "description": "This app is super cool",
    "id": 4,
    "name": "Secure Image",
    "updatedAt": "2020-04-28T00:00:00.000Z",
  },
]
`;

exports[`Profile event handlers Fetch all profiles should throw 1`] = `"Unable fetch all project profiles"`;

exports[`Profile event handlers Fetch single profile should throw 1`] = `"Unable fetch project profile with ID 1"`;
